pkgs/extension_discovery/lib/extension_discovery.dart

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// Find extensions with [findExtensions].
+library;
+
+import 'dart:collection' show UnmodifiableListView;
+import 'dart:convert' show jsonDecode;
+import 'dart:io' show File, IOException;
+import 'dart:isolate' show Isolate;
+
+import 'src/io.dart';
+import 'src/package_config.dart';
+import 'src/registry.dart';
+
+export 'src/package_config.dart' show PackageConfigException;
+
+/// Information about an extension for target package.
+final class Extension {
+  /// Name of the package providing an extension.
+  final String package;
+
+  /// Absolute path to the package root.
+  ///
+  /// This folder usually contains a `lib/` folder and a `pubspec.yaml`
+  /// (assuming dependencies are fetched using the pub package manager).
+  final Uri rootUri;
+
+  /// Path to the library import path relative to [rootUri].
+  ///
+  /// In Dart code the `package:<package>/<path>` will be resolved as
+  /// `<rootUri>/<packageUri>/<path>`.
+  ///
+  /// This is almost always `lib/`, and technically, it's probably perf
+  final Uri packageUri;
+
+  /// Contents of `extension/<targetPackage>/config.json` parsed as JSON.
+  ///
+  /// If parsing JSON from this file failed, then no [Extension] entry
+  /// will exist.
+  ///
+  /// This field is always a structure consisting of the following types:
+  ///  * `null`,
+  ///  * [bool] (`true` or `false`),
+  ///  * [String],
+  ///  * [num] ([int] or [double]),
+  ///  * [List<Object?>], and,
+  ///  * [Map<String, Object?>].
+  final Object? config;
```

Comment:
if this is parsed json, why not make the type `Map<String, Object?>`?

Reply:
You can technically write:
```
null
```

in you `config.json`.

-----------------

I don't think it makes sense to use something other than `Map` as root, but on the other hand I see no reason why this package should care about what the contents of `config.json` is -- just that parses as JSON.